### PR TITLE
HAAR-885:  Get user roles by username

### PR DESCRIPTION
**Pull request by ravmoj**
Get user roles by username

### CODE REVIEW

**Comment by ravmoj** [comment]
Currently moving /api/user/{username}/roles endpoint. This endpoint is used by PCMS and Pathfinder services
1.For Nomis Users roles: getting an additional role : PRISON
2.For Delius users : getting an additional role : PROBATION
These are are manually added in [uk.gov](http://uk.gov/).justice.digital.hmpps.oauth2server.delius.model.DeliusUserPersonDetails#getAuthorities and [uk.gov](http://uk.gov/).justice.digital.hmpps.oauth2server.nomis.model.NomisUserPersonDetails#getAuthorities
@brightonsbox , @SimonMitchellMOJ , @karenmillermoj

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/AuthExceptionHandler.kt [hunk 1]
```diff
@@ -12,8 +10,7 @@ import uk.gov.justice.digital.hmpps.manageusersapi.service.auth.NotFoundExceptio
 
 // Exception Handler to mirror the exceptions thrown by Auth.  Needed to ensure all endpoints moved from auth into this
 // service return the same exception detail.
-@RestControllerAdvice
-@Order(HIGHEST_PRECEDENCE)
+@RestControllerAdvice()
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/AuthExceptionHandler.kt:13]: Possible merge issue ?  We still need precedence

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt [hunk 2]
```diff
@@ -1956,6 +1956,28 @@ class ExternalUsersApiMockServer : WireMockServer(WIREMOCK_PORT) {
...
+                  "errorCode": null,
+                  "userMessage": "User not found: Account for username $username not found",
+                  "developerMessage": "Account for username $username not found",
+                  "moreInfo": null
+                } 
+              """.trimIndent()
+            )
+        )
+    )
+  }
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt:1979]: we already have stubNoUsersFoundForUsername - maybe we should combine these and pass in the url?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/UserServiceTest.kt [hunk 3]
```diff
@@ -113,6 +118,67 @@ class UserServiceTest {
...
+        verifyNoInteractions(deliusUserApiService)
+      }
+
+      @Test
+      fun `find roles of nomis user`() {
+        whenever(externalRolesApiService.findRolesByUsernameOrNull(anyString())).thenReturn(null)
+        whenever(nomisUserApiService.findUserByUsername(anyString())).thenReturn(createNomisUser())
+
+        val userRoleList = userService.findRolesByUsername("nuser_gen")
+        println(userRoleList)
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/UserServiceTest.kt:166]: prob don't need the println

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt (hunk 2 above)
Comment (karenmillermoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt:1979]: have added this query to a separate ticket HAAR-1223

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/UserControllerIntTest.kt [hunk 4]
```diff
@@ -296,4 +297,98 @@ class UserControllerIntTest : IntegrationTestBase() {
...
+        .jsonPath("[*].roleCode").value<List<String>> {
+          assertThat(it).contains("MAINTAIN_ACCESS_ROLES")
+          assertThat(it).contains("GLOBAL_SEARCH")
+          assertThat(it).contains("HMPPS_REGISTERS_MAINTAINER")
+          assertThat(it).contains("HPA_USER")
+        }
+    }
+
+    @Test
+    fun `should fail with not_found for invalid user`() {
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/UserControllerIntTest.kt:339]: Is this a duplicate?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/AuthExceptionHandler.kt (hunk 1 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/AuthExceptionHandler.kt]: Updated

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt (hunk 2 above)
Comment (ravmoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/ExternalUsersApiMockServer.kt]: Updated this stub

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/UserServiceTest.kt (hunk 3 above)
Comment (ravmoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/UserServiceTest.kt]: removed

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/UserControllerIntTest.kt (hunk 4 above)
Comment (ravmoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/UserControllerIntTest.kt]: removed

**Review by karenmillermoj — APPROVED** [review]
lgtm

**Pull request merged**